File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/query/DefaultRequeryEntityMetadata.java
package com.coupang.springframework.data.requery.repository.query;

import io.requery.Entity;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.core.annotation.AnnotatedElementUtils;
import org.springframework.util.Assert;
import org.springframework.util.StringUtils;

/**
 * Default implementation for {@link RequeryEntityMetadata}.
 *
 * @author debop@coupang.com
 * @since 18. 6. 7
 */
@Slf4j
public class DefaultRequeryEntityMetadata<T> implements RequeryEntityMetadata<T> {

    @NotNull
    public static <T> DefaultRequeryEntityMetadata<T> of(@NotNull Class<T> domainClass) {
        Assert.notNull(domainClass, "domainClass must not be null");
        return new DefaultRequeryEntityMetadata<>(domainClass);
    }

    private final Class<T> domainClass;

    public DefaultRequeryEntityMetadata(@NotNull Class<T> domainClass) {
        Assert.notNull(domainClass, "domainClass must not be null!");
        this.domainClass = domainClass;
    }

    @Override
    public String getEntityName() {
        Entity entity = AnnotatedElementUtils.findMergedAnnotation(domainClass, Entity.class);

        log.debug("Get entity name... domainClass={}, entity={}", domainClass.getName(), entity);

        return (entity != null) && StringUtils.hasText(entity.name())
               ? entity.name()
               : domainClass.getSimpleName();
    }

    @Override
    public String getModelName() {
        Entity entity = AnnotatedElementUtils.findMergedAnnotation(domainClass, Entity.class);

        log.debug("Get model name... domainClass={}, entity={}", domainClass.getName(), entity);

        return (entity != null) && StringUtils.hasText(entity.model())
               ? entity.model()
               : "default";
    }

    @Override
    public Class<T> getJavaType() {
        return domainClass;
    }
}

File: spring-data-requery/src/main/kotlin/org/springframework/data/requery/repository/query/DefaultRequeryEntityMetadata.kt
package org.springframework.data.requery.repository.query

import io.requery.Entity
import mu.KotlinLogging
import org.springframework.core.annotation.AnnotatedElementUtils

/**
 * DefaultRequeryEntityMetadata
 *
 * @author debop@coupang.com
 */
class DefaultRequeryEntityMetadata<E>(val domainClass: Class<E>): RequeryEntityMetadata<E> {

    companion object {
        private val log = KotlinLogging.logger { }

        private const val DEFAULT_MODEL_NAME = "default"

        @JvmStatic
        fun <E> of(domainClass: Class<E>): DefaultRequeryEntityMetadata<E> =
            DefaultRequeryEntityMetadata(domainClass)
    }

    private val entity: Entity? by lazy {
        AnnotatedElementUtils.findMergedAnnotation(domainClass, io.requery.Entity::class.java)
    }

    override val entityName: String by lazy {
        log.trace { "Get entity name ... domainClass=$domainClass, entity=$entity" }

        val entity = this.entity
        when {
            entity != null && entity.name.isNotBlank() -> entity.name
            else -> domainClass.simpleName
        }
    }


    override val modelName: String by lazy {
        log.trace { "Get model name ... domainClass=$domainClass, entity=$entity" }

        entity?.let {
            if(it.model.isNotBlank()) it.model else DEFAULT_MODEL_NAME
        } ?: ""
    }

    override fun getJavaType(): Class<E> = domainClass

}